cpp/ql/src/semmle/code/cpp/Element.qll

```diff
@@ -2,13 +2,31 @@ import semmle.code.cpp.Location
 private import semmle.code.cpp.Enclosing
 private import semmle.code.cpp.internal.Type
 
+cached @element resolveElement(@element e) {
+  if isClass(e)
+  then result = resolve(e)
+  else result = e
+}
+
+cached Element mkElement(@element e) {
+  result = resolveElement(e)
+}
+
+cached @element unresolveElement(Element e) {
+  resolveElement(result) = e
+}
+
+cached @element underlyingElement(Element e) {
+  result = e
+}
```

Comment:
Please add QLDoc to these four predicates to explain how they are supposed to be used. New code might easily get these calls wrong, especially if we don't all understand how to use them. In particular, how do I choose between using `unresolveElement` and `underlyingElement`?

Reply:
Done